docs/changelog/3_x.rst

```diff
@@ -0,0 +1,729 @@
+      default::City {name: 'Phoenix'},
+      default::City {name: 'Philadelphia'},
+      default::City {name: 'Dallas'}
+    }
+
+
+``assert`` function
+-------------------
+
+The new ``assert`` function lets you do handy things like create powerful
+constraints when paired with triggers:
+
+.. code-block:: sdl
+
+      type Person {
+        required name: str;
+        multi friends: User;
+        multi enemies: User;
+
+        trigger prohibit_frenemies after insert, update for each do (
+          assert(
+            not exists (__new__.friends intersect __new__.enemies),
+            message := "Invalid frenemies",
+          )
+        )
+      }
+
+.. code-block:: edgeql-repl
+
+    db> insert Person {name := 'Quincey Morris'};
+    {default::Person {id: e4a55480-d2de-11ed-93bd-9f4224fc73af}}
+    db> insert Person {name := 'Dracula'};
+    {default::Person {id: e7f2cff0-d2de-11ed-93bd-279780478afb}}
+    db> update User
+    ... filter .name = 'Quincey Morris'
+    ... set {
+    ...   enemies := (select Person filter .name = 'Dracula')
+    ... };
+    {default::Person {id: e4a55480-d2de-11ed-93bd-9f4224fc73af}}
+    db> update User
+    ... filter .name = 'Quincey Morris'
+    ... set {
+    ...   friends := (select Person filter .name = 'Dracula')
+    ... };
+    edgedb error: EdgeDBError: Invalid frenemies
+
+You can use it in other contexts too — any time you want to throw an error when
+things don't go as planned.
+
+
+Additional changes
+==================
+
+EdgeQL
+------
+
+* Support custom user-defined error messages for access policies
+  (:eql:gh:`#4529`)
+
+  .. code-block:: sdl
+
+    type User {
+      required property email -> str { constraint exclusive; };
+      required property is_admin -> bool { default := false };
+      access policy admin_only
+        allow all
+        using (global current_user.is_admin ?? false) {
+          errmessage := 'Only admins may query Users'
+        };
+    }
+
+* Support casting a UUID to a type (:eql:gh:`#4469`). This is a handy way to
+  select an object, assuming the type you cast into has an object with the UUID
+  being cast.
+
+  .. code-block:: edgeql-repl
+
+      db> select <Hero><uuid>'01d9cc22-b776-11ed-8bef-73f84c7e91e7';
+      {default::Hero {id: 01d9cc22-b776-11ed-8bef-73f84c7e91e7}}
+
+* Add the :eql:func:`json_object_pack` function to construct JSON from an array
+  of key/value tuples. (:eql:gh:`#4474`)
+
+  .. code-block:: edgeql-repl
+
+      db> select json_object_pack({("hello", <json>"world")});
+      {Json("{\"hello\": \"world\"}")}
+
+* Support tuples as query arguments (:eql:gh:`#4489`)
+
+  .. code-block:: edgeql
+
+      select <tuple<str, bool>>$var;
+      select <optional tuple<str, bool>>$var;
+      select <tuple<name: str, flag: bool>>$var;
+      select <optional tuple<name: str, flag: bool>>$var;
+
+* Add the syntax for abstract indexes (:eql:gh:`#4691`)
+
+  Exposes some Postgres indexes that you can use in your schemas. These are
```

Comment:
could link to the Postgres docs here for people who want more details about the fancy index types:

https://www.postgresql.org/docs/current/indexes-types.html